### PR TITLE
cardano-cli: Add option `--stake-address` to several subcommands

### LINKED ISSUE #2723: [FR] - add option to stake-address registration-certificate to take an address
**Internal/External**
*Internal*   if an IOHK staff member.

**Area**
*Token Locking*    Related to Token Locking (Allegra).
*Native Tokens*    Related to Native Tokens (Mary).
*Other*   Any other topic (Delegation, Ranking, ...).

**Describe the feature you'd like**
Currently it takes a stake vkey file or a stake vkey as text, but not a stake address. But the resulting certificate is just the hash of the stake vkey (e.g. address without the first nibble). Would be nice to have an additional --stake-address param to handle that case.

**Describe alternatives you've considered**
I wrote a python script to write my own stake registration cert files

**Additional context / screenshots**

### PR DESCRIPTION
**Pull request by cblp**
Fix #2723

### CODE REVIEW

**Comment by cblp** [comment]
```
Usage: cardano-cli stake-address registration-certificate 
            ( --stake-verification-key STRING
            | --stake-verification-key-file FILE
            | --stake-script-file FILE
            | --stake-address ADDRESS
            )
            --out-file FILE

  Create a stake address registration certificate

Available options:
  --stake-verification-key STRING
                           Stake verification key (Bech32 or hex-encoded).
  --stake-verification-key-file FILE
                           Filepath of the staking verification key.
  --stake-script-file FILE Filepath of the staking script.
  --stake-address ADDRESS  Target stake address (bech32 format).
  --out-file FILE          The output file.
  -h,--help                Show this help text
```

@disassembler is this what you expected?

File: cardano-cli/src/Cardano/CLI/Shelley/Commands.hs [hunk 1]
```diff
@@ -118,7 +118,7 @@ data StakeAddressCmd
   = StakeAddressKeyGen VerificationKeyFile SigningKeyFile
   | StakeAddressKeyHash (VerificationKeyOrFile StakeKey) (Maybe OutputFile)
   | StakeAddressBuild StakeVerifier NetworkId (Maybe OutputFile)
-  | StakeRegistrationCert StakeVerifier OutputFile
+  | StakeRegistrationCert (Either StakeVerifier StakeAddress) OutputFile
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Commands.hs:121]: A more seamless change would be to add an additional constructor to the `StakeVerifier` data type that allows  the specification of a `StakeAddress`.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
See comment

File: cardano-cli/src/Cardano/CLI/Shelley/Key.hs [hunk 2]
```diff
@@ -392,6 +392,7 @@ data PaymentVerifier
 data StakeVerifier
   = StakeVerifierKey (VerificationKeyOrFile StakeKey)
   | StakeVerifierScriptFile ScriptFile
+  | StakeVerifierAddress StakeAddress
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Key.hs:395]: :+1: This is much cleaner now.

Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Key.hs:395]: I was not sure all components using StakeVerifier will accept StakeAddress, so I started with ad hoc solution, and then assured I can modify this type.

File: cardano-cli/src/Cardano/CLI/Shelley/Commands.hs (hunk 1 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Commands.hs]: done

**Comment by cblp** [comment]
@Jimbo4350 does this CLI change require updating documentation or version bump?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs [hunk 3]
```diff
@@ -88,51 +90,24 @@ runStakeAddressBuild
...
-      stakeVerKey <- firstExceptT ShelleyStakeAddressCmdReadKeyFileError
-        . newExceptT
-        $ readVerificationKeyOrFile AsStakeKey stakeVerKeyOrFile
-      let stakeCred = StakeCredentialByKey (verificationKeyHash stakeVerKey)
-      writeRegistrationCert stakeCred
+runStakeCredentialRegistrationCert stakeVerifier (OutputFile oFp) = do
+  stakeCred <- loadStakeCredentialFromVerifier stakeVerifier
+  writeRegistrationCert stakeCred
  where
+
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:110]: Unnecessary space

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs [hunk 4]
```diff
@@ -216,3 +172,45 @@ runStakeCredentialDeRegistrationCert stakeVerifier (OutputFile oFp) =
...
+loadStakeCredentialFromVerifier
+  :: StakeVerifier -> ExceptT ShelleyStakeAddressCmdError IO StakeCredential
+
+loadStakeAddressFromVerifier
+  :: NetworkId
+  -> StakeVerifier
+  -> ExceptT ShelleyStakeAddressCmdError IO StakeAddress
+
+(loadStakeCredentialFromVerifier, loadStakeAddressFromVerifier) =
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:185]: This is quite a confusing way to define these functions. Let's separate them. 

I am curious though, why did you define them like this?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs (hunk 4 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:185]: They have common internal logic that should be hidden.

Alternative: define 3 functions, but the 3rd one will be visible to all other functions in the module.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Address.hs [hunk 5]
```diff
@@ -232,4 +235,3 @@ runAddressBuildScript
 runAddressBuildScript scriptFile networkId mOutputFile = do
   liftIO $ deprecationWarning "'address build'"
   runAddressBuild (PaymentVerifierScriptFile scriptFile) Nothing networkId mOutputFile
-
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Address.hs:235]: Please undo this deletion

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs (hunk 4 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:185]: Yes please define three separate functions.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:110]: If you look at the rest of the code base, we don't put a space after `where`. Please conform to the existing codestyle.

Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:109]: okay, formatted similar to other functions in this file

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs [hunk 6]
```diff
@@ -216,3 +172,44 @@ runStakeCredentialDeRegistrationCert stakeVerifier (OutputFile oFp) =
...
+-- which one is closer.
+--
+-- For internal use in 'loadStakeCredentialFromVerifier' and
+-- 'loadStakeAddressFromVerifier' only
+loadStakeVerifier
+  :: StakeVerifier
+  -> ExceptT
+        ShelleyStakeAddressCmdError
+        IO
+        (Either StakeAddress StakeCredential)
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:199]: `runStakeAddressBuild` needs full address. Dou you propose dropping network information here and append it back in `runStakeAddressBuild`?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs [hunk 7]
```diff
@@ -216,3 +172,44 @@ runStakeCredentialDeRegistrationCert stakeVerifier (OutputFile oFp) =
 
     deregCertDesc :: TextEnvelopeDescr
     deregCertDesc = "Stake Address Deregistration Certificate"
+
+
+loadStakeCredentialFromVerifier
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs]: done

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs]: done; found a simpler solution

File: cardano-api/src/Cardano/Api/Address.hs [hunk 8]
```diff
@@ -649,3 +650,6 @@ fromShelleyStakeReference (Shelley.StakeRefPtr ptr) =
 fromShelleyStakeReference Shelley.StakeRefNull =
   NoStakeAddress
 
+-- | Get credential from a stake address. This drops the network information.
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/Address.hs]: done

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs [hunk 9]
```diff
@@ -216,3 +172,31 @@ runStakeCredentialDeRegistrationCert stakeVerifier (OutputFile oFp) =
 
     deregCertDesc :: TextEnvelopeDescr
     deregCertDesc = "Stake Address Deregistration Certificate"
+
+
+getStakeCredentialFromVerifier
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:177]: Great. This is much simpler now.

**Review by Jimbo4350 — APPROVED** [review]
:+1:

**Comment by newhoggy** [comment]
This affects the following commands:

* `address build`
* `stake-address build`
* `stake-address registration-certificate`
* `stake-address delegation-certificate`
* `stake-address deregistration-certificate`

**Pull request merged**